src/weekly/mod.rs

```diff
@@ -0,0 +1,411 @@
+
+/// Calendar for periodic time-windows, recurring on weekly basis.
+#[derive(Clone, Debug)]
+pub struct WeeklyCalendar {
+    /// An immutable set of (possibly overlapping) intervals.
+    windows: IntervalTree<MinuteInWeek, WeeklyWindow>,
+}
+
+impl WeeklyCalendar {
+    /// Create a calendar from a vector of weekly windows.
+    pub fn new(input: Vec<WeeklyWindow>) -> Self {
+        let intervals = input
+            .into_iter()
+            .map(|win| Element::from((win.range_weekly_minutes(), win)));
+
+        Self {
+            windows: IntervalTree::from_iter(intervals),
+        }
+    }
+
+    /// Return whether datetime is contained in this weekly calendar.
+    pub fn contains_datetime(&self, datetime: &DateTime<Utc>) -> bool {
+        let timepoint = utils::datetime_as_weekly_minute(datetime);
+        self.windows.query_point(timepoint).count() > 0
+    }
+
+    /// Return the measured length of the calendar.
+    ///
+    /// In case of overlapping windows, measured length is the actual amount
+    /// of weekly minutes in the calendar. Overlapped intervals are coalesced
+    /// in order to avoid double-counting.
+    pub fn length_minutes(&self) -> u64 {
+        let mut measured = 0u32;
+        let mut last_range = Range {
+            start: 0u32,
+            end: 0u32,
+        };
+
+        for win in self.windows.iter_sorted() {
+            if win.range.start > last_range.end {
+                // Non-overlapping window, update accumulator and use this as last range.
+                let last_length = last_range.end.saturating_sub(last_range.start + 1);
+                measured = measured.saturating_add(last_length);
+                last_range = win.range.clone();
+            } else {
+                // Overlapping window, coalesce into the last range.
+                last_range.end = u32::max(last_range.end, win.range.end);
+            };
+        }
+        // Account for the still pending length of the last range.
+        let last_length = last_range.end.saturating_sub(last_range.start + 1);
+        measured = measured.saturating_add(last_length);
+
+        u64::from(measured)
+    }
+
+    /// Return total length of all windows in the calendar.
+    ///
+    /// In case of overlapping windows, total length can be larger than the
+    /// actual amount of weekly minutes in the calendar.
+    #[cfg(test)]
+    pub fn total_length_minutes(&self) -> u64 {
+        self.windows.iter().fold(0u64, |len, win| {
+            len.saturating_add(win.value.length_minutes().into())
+        })
+    }
+
+    /// Return all weekly windows (if any) which contain a given datetime.
+    #[cfg(test)]
+    pub fn containing_windows(&self, datetime: &DateTime<Utc>) -> Vec<&WeeklyWindow> {
+        let timepoint = utils::datetime_as_weekly_minute(datetime);
+        self.windows
+            .query_point(timepoint)
+            .map(|elem| &elem.value)
+            .collect()
+    }
+}
+
+impl Serialize for WeeklyCalendar {
```

Comment:
Oh interesting. Why do we need a serialization support for this?

Reply:
We do embed this in the final runtime configuration, whose top-level container has a `Serialize` annotation.

I don't think it's used right now. I used to have some private logic for runtime config introspection, but I didn't turn into a public API as I wasn't unsure about the details. I can try resurrecting that if it seems useful.
Keeping the datastructure serializable is a low incremental effort and unlocks such future usage.